compiler/rustc_codegen_ssa/src/back/link.rs

```diff
@@ -1986,11 +1986,11 @@ fn linker_with_args<'a>(
         tmpdir,
     );
 
-    // Upstream dynamic native libraries linked with `#[link]` attributes at and `-l`
-    // command line options.
-    // If -Zlink-native-libraries=false is set, then the assumption is that an
-    // external build system already has the native dependencies defined, and it
-    // will provide them to the linker itself.
+    // Dynamic native libraries from upstream crates.
+    //
+    // FIXME: Merge this to `add_upstream_rust_crates` so that all native libraries are linked
+    // together with their respective upstream crates, and in their originally specified order.
+    // This may be slightly breaking due to our use of `--as-needed` and needs a crater run.
```

Comment:
I think this would break mustang if the libc crate happens to be included before mustang as in that case it wouldn't be possible to override libc functions anymore I think. The change is also not all that useful as lld and mold only care about link order for symbol precedence and don't deny backward references like ld.bfd and gold.

Reply:
I have a few motivations for this change:
1. Preserving the user-specified argument order (https://github.com/rust-lang/rfcs/blob/master/text/2951-native-link-modifiers.md#relative-order-of--l-and--clink-args-options, https://github.com/rust-lang/rust/issues/99427)
2. Consistent behavior between libraries linked from the current crate and dependency crates - note how everything for the current crate is currently passed together and in order (for `1.`), and before dependency crates (for `3.`).
3. Guarding against accidental dependencies on native libraries (well, maybe not with lld and mold as you said, but with more traditional linkers). The mustang regression (if it happens) likely falls into this category, if it breaks then some kind of undeclared dependency is the root cause.

Reply:
@sunfishcode WDYT Would this break mustang or not?

Reply:
Mustang currently works by creating an empty libc.a, such that it doesn't care where in the link order -lc appears.

Reply:
I see.